Automated Smoketest: :x:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-123.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.22.1,  cUrl: 7.67.0,  node.js: v16.13.1,  npm: 8.1.2
Sha: de752e50bc04babbc0ae4a8b1a58b914dd31874c
Containerized:False
Number of Error(s): 26
CMake Error at cmake_modules/commonSetup.cmake:581 (add_subdirectory):
CMake Error at cmake_modules/commonSetup.cmake:581 (add_subdirectory):
CMake Error at cmake_modules/commonSetup.cmake:581 (add_subdirectory):
CMake Error at cmake_modules/commonSetup.cmake:581 (add_subdirectory):
CMake Error at cmake_modules/commonSetup.cmake:581 (add_subdirectory):
CMake Error at cmake_modules/commonSetup.cmake:581 (add_subdirectory):
CMake Error at cmake_modules/commonSetup.cmake:581 (add_subdirectory):
CMake Error at cmake_modules/commonSetup.cmake:581 (add_subdirectory):
CMake Error at cmake_modules/commonSetup.cmake:581 (add_subdirectory):
CMake Error at cmake_modules/commonSetup.cmake:581 (add_subdirectory):
CMake Error at cmake_modules/commonSetup.cmake:581 (add_subdirectory):
CMake Error at cmake_modules/commonSetup.cmake:581 (add_subdirectory):
CMake Error at cmake_modules/commonSetup.cmake:581 (add_subdirectory):
CMake Error at cmake_modules/commonSetup.cmake:581 (add_subdirectory):
CMake Error at cmake_modules/commonSetup.cmake:581 (add_subdirectory):
CMake Error at cmake_modules/commonSetup.cmake:581 (add_subdirectory):
CMake Error at cmake_modules/commonSetup.cmake:581 (add_subdirectory):
CMake Error at cmake_modules/commonSetup.cmake:581 (add_subdirectory):
CMake Error: File /home/centos/smoketest/PR-15671/HPCC-Platform/build-config.h.cmake does not exist.
CMake Error at CMakeLists.txt:369 (configure_file):
CMake Error at CMakeLists.txt:447 (include):
CMake Error at CMakeLists.txt:517 (add_subdirectory):
CMake Error: File /home/centos/smoketest/PR-15671/HPCC-Platform/cmake_uninstall.cmake.in does not exist.
CMake Error at CMakeLists.txt:520 (configure_file):
-- Configuring incomplete, errors occurred!
make: getcwd: No such file or directory
CMake Error at cmake_modules/commonSetup.cmake:581 (add_subdirectory):
  add_subdirectory given source tools which is not an existing directory.
Call Stack (most recent call first):
  CMakeLists.txt:173 (HPCC_ADD_SUBDIRECTORY)
CMake Error at cmake_modules/commonSetup.cmake:581 (add_subdirectory):
  add_subdirectory given source common which is not an existing directory.
Call Stack (most recent call first):
  CMakeLists.txt:174 (HPCC_ADD_SUBDIRECTORY)
CMake Error at cmake_modules/commonSetup.cmake:581 (add_subdirectory):
  add_subdirectory given source dali which is not an existing directory.
Call Stack (most recent call first):
  CMakeLists.txt:175 (HPCC_ADD_SUBDIRECTORY)
CMake Error at cmake_modules/commonSetup.cmake:581 (add_subdirectory):
  add_subdirectory given source ecl which is not an existing directory.
Call Stack (most recent call first):
  CMakeLists.txt:176 (HPCC_ADD_SUBDIRECTORY)
CMake Error at cmake_modules/commonSetup.cmake:581 (add_subdirectory):
  add_subdirectory given source ecllibrary which is not an existing
  directory.
Call Stack (most recent call first):
  CMakeLists.txt:177 (HPCC_ADD_SUBDIRECTORY)
CMake Error at cmake_modules/commonSetup.cmake:581 (add_subdirectory):
  add_subdirectory given source esp which is not an existing directory.
Call Stack (most recent call first):
  CMakeLists.txt:178 (HPCC_ADD_SUBDIRECTORY)
CMake Error at cmake_modules/commonSetup.cmake:581 (add_subdirectory):
  add_subdirectory given source fs which is not an existing directory.
Call Stack (most recent call first):
  CMakeLists.txt:179 (HPCC_ADD_SUBDIRECTORY)
CMake Error at cmake_modules/commonSetup.cmake:581 (add_subdirectory):
  add_subdirectory given source plugins which is n
 ... ( comment is too long, 197 lines, 10836 bytes, truncated) ... 
[Build error:](https://gist.github.com/32234ee93dccee55d653f7e6ef625bb9)
Build log [Build log](https://gist.github.com/HPCCSmoketest/32234ee93dccee55d653f7e6ef625bb9#file-result-22-01-20-10-05-39-log)

```

--- a/ecl/hql/hqlgram.hpp
+++ b/ecl/hql/hqlgram.hpp
@@ -1254,7 +1254,7 @@ class HqlLex
         void pushMacro(IHqlExpression *expr);
         void pushText(IFileContents * text, int startLineNo, int startColumn);
         void pushText(const char *s, int startLineNo, int startColumn);
-        bool getParameter(StringBuffer &curParam, const char* for_what);
+        bool getParameter(StringBuffer &curParam, const char* directive, const ECLlocation & location);
         IValue *foldConstExpression(const ECLlocation & errpos, IHqlExpression * expr, IXmlScope *xmlScope);
         IValue *parseConstExpression(const ECLlocation & errpos, StringBuffer &curParam, IXmlScope *xmlScope);
         IValue *parseConstExpression(const ECLlocation & errpos, IFileContents * contents, IXmlScope *xmlScope);
@@ -1286,7 +1286,7 @@ class HqlLex
         void doInModule(attribute & returnToken);
         void doMangle(attribute & returnToken, bool de);
         void doUniqueName(attribute & returnToken);
-        void doSkipUntilEnd(attribute & returnToken, const char * forwhat);
+        void doSkipUntilEnd(attribute & returnToken, const char* directive, const ECLlocation & location);
 
         void processEncrypted();
         void checkSignature(const attribute & dummyToken);
@@ -1294,7 +1294,7 @@ class HqlLex
         void declareUniqueName(const char* name, const char * pattern);
         void checkNextLoop(bool first);
 
-        bool getDefinedParameter(StringBuffer &curParam, attribute & returnToken, const char* for_what, SharedHqlExpr & resolved);
+        bool getDefinedParameter(StringBuffer &curParam, attribute & returnToken, const char* directive, const ECLlocation & location, SharedHqlExpr & resolved);
 
         int processStringLiteral(attribute & returnToken, char *CUR_TOKEN_TEXT, unsigned CUR_TOKEN_LENGTH, int oldColumn, int oldPosition);
 

--- a/ecl/hql/hqlparse.cpp
+++ b/ecl/hql/hqlparse.cpp
@@ -620,7 +620,7 @@ void HqlLex::processEncrypted()
 }
 
 /* Return: true if more parameter(s) left. */
-bool HqlLex::getParameter(StringBuffer &curParam, const char* for_what)
+bool HqlLex::getParameter(StringBuffer &curParam, const char* directive, const ECLlocation & location)
 {
     unsigned parenDepth = 1;
     attribute nextToken;
@@ -649,9 +649,7 @@ bool HqlLex::getParameter(StringBuffer &curParam, const char* for_what)
             break;
         case EOF:
             {
-                StringBuffer msg("EOF encountered while gathering parameters for ");
-                msg.append(for_what);
-                reportError(nextToken, ERR_TMPLT_EOFINPARAM, "%s", msg.str());
+                reportError(location, ERR_TMPLT_EOFINPARAM, "EOF encountered while gathering parameters for %s", directive);
             }
             return false;
         default:
@@ -663,7 +661,7 @@ bool HqlLex::getParameter(StringBuffer &curParam, const char* for_what)
     }
 }
 
-void HqlLex::doSkipUntilEnd(attribute & returnToken, const char * forwhat)
+void HqlLex::doSkipUntilEnd(attribute & returnToken, const char* directive, const ECLlocation & location)
 {
     while (skipNesting)
     {
@@ -671,9 +669,7 @@ void HqlLex::doSkipUntilEnd(attribute & returnToken, const char * forwhat)
         returnToken.release();
         if (tok == EOF)
         {
-            StringBuffer msg;
-            msg.appendf("Unexpected EOF in %s: #END expected",forwhat);
-            reportError(returnToken, ERR_TMPLT_HASHENDEXPECTED, "%s", msg.str());
+            reportError(location, ERR_TMPLT_HASHENDEXPECTED, "Unexpected EOF in %s: #END expected", directive);
             clearNestedHash();      // prevent unnecessary more error messages
             break;
         }
@@ -683,19 +679,18 @@ void HqlLex::doSkipUntilEnd(attribute & returnToken, const char * forwhat)
 void HqlLex::doIf(attribute & returnToken, bool isElseIf)
 {
     ECLlocation location = returnToken.pos;
-    StringBuffer forwhat;
-    forwhat.appendf("#IF(%d,%d)",returnToken.pos.lineno,returnToken.pos.column);
+    const char * directive = "#IF";
 
     int tok = yyLex(returnToken, LEXnone, 0);
     if (tok != '(')
         reportError(returnToken, ERR_EXPECTED_LEFTCURLY, "( expected"); // MORE - make it fatal!
 
     StringBuffer curParam("(");
-    if (getParameter(curParam, forwhat.str()))
+    if (getParameter(curParam, directive, location))
     {
         reportError(returnToken, ERR_OPERANDS_TOOMANY, "too many operands");
         StringBuffer dummy;
-        while (getParameter(dummy,forwhat.str()))
+        while (getParameter(dummy, directive, location))
             ;
     }
     curParam.append(')');
@@ -705,7 +700,7 @@ void HqlLex::doIf(attribute & returnToken, bool isElseIf)
         setHashEndFlags(0);
         skipNesting = 1;
         if (!isElseIf)
-            doSkipUntilEnd(returnToken, forwhat);
+            doSkipUntilEnd(returnToken, directive, location);
     }
     else
         setHashEndFlags(HEFhadtrue);
@@ -713,8 +708,8 @@ void HqlLex::doIf(attribute & returnToken, bool isElseIf)
 
 int HqlLex::doElse(attribute & returnToken, LexerFlags lookupFlags, const short * activeState, bool isElseIf)
 {
-    StringBuffer forwhat;
-    forwhat.appendf("#%s(%d,%d)",isElseIf ? "ELSEIF" : "ELSE", returnToken.pos.lineno,returnToken.pos.column);
+    ECLlocation location = returnToken.pos;
+    const char * directive = isElseIf ? "#ELSEIF" : "#ELSE";
 
     if ((hashendKinds.ordinality() == 0) || (hashendKinds.tos() != HashStmtIf))
     {
@@ -734,7 +729,7 @@ int HqlLex::doElse(attribute & returnToken, LexerFlags lookupFlags, const short 
     {
     case 0:
         skipNesting = 1;
-        doSkipUntilEnd(returnToken, forwhat);
+        doSkipUntilEnd(returnToken, directive, location);
         return yyLex(returnToken, lookupFlags, activeState);
     case 1:
         if (flags & HEFhadtrue)
@@ -774,9 +769,7 @@ int HqlLex::doEnd(attribute & returnToken, LexerFlags lookupFlags, const short *
 
 void HqlLex::doDeclare(attribute & returnToken)
 {
-    StringBuffer forwhat;
-    forwhat.appendf("#DECLARE(%d,%d)",returnToken.pos.lineno,returnToken.pos.column);
-
+    ECLlocation location = returnToken.pos;
     IIdAtom * name = NULL;
     if (yyLex(returnToken, LEXnone, 0) != '(')
     {
@@ -790,9 +783,7 @@ void HqlLex::doDeclare(attribute & returnToken)
         int tok = yyLex(returnToken, LEXidentifier, 0);
         if (tok == EOF)
         {
-            StringBuffer msg;
-            msg.append("Unexpected EOF in ").append(forwhat.str()).append(": ')' expected");
-            reportError(returnToken, ERR_TMPLT_HASHENDEXPECTED, "%s", msg.str());
+            reportError(location, ERR_TMPLT_HASHENDEXPECTED, "Unexpected EOF in #DECLARE: ')' expected");
             clearNestedHash();      // prevent unnecessary more error messages
             return;
         }
@@ -814,9 +805,7 @@ void HqlLex::doDeclare(attribute & returnToken)
             continue;
         else if (tok == EOF)
         {
-            StringBuffer msg;
-            msg.append("Unexpected EOF in ").append(forwhat.str()).append(": ) expected");
-            reportError(returnToken, ERR_TMPLT_HASHENDEXPECTED, "%s", msg.str());
+            reportError(location, ERR_TMPLT_HASHENDEXPECTED, "Unexpected EOF in #DECLARE: ')' expected");
             clearNestedHash();      // prevent unnecessary more error messages
         }
         else
@@ -829,8 +818,8 @@ void HqlLex::doDeclare(attribute & returnToken)
 
 void HqlLex::doExpand(attribute & returnToken)
 {
-    StringBuffer forwhat;
-    forwhat.appendf("#DECLARE(%d,%d)",returnToken.pos.lineno,returnToken.pos.column);
+    ECLlocation location = returnToken.pos;
+    const char * directive = "#EXPAND";
 
     if (yyLex(returnToken, LEXnone, 0) != '(')
     {
@@ -840,11 +829,11 @@ void HqlLex::doExpand(attribute & returnToken)
     }
 
     StringBuffer curParam("(");
-    if (getParameter(curParam,forwhat.str()))
+    if (getParameter(curParam, directive, location))
     {
         reportError(returnToken, ERR_OPERANDS_TOOMANY, "Too many operands");
         StringBuffer dummy;
-        while (getParameter(dummy,forwhat.str()))
+        while (getParameter(dummy, directive, location))
             ;
     }
     curParam.append(')');
@@ -860,8 +849,8 @@ void HqlLex::doExpand(attribute & returnToken)
 
 void HqlLex::doSet(attribute & returnToken, bool append)
 {
-    StringBuffer forwhat;
-    forwhat.appendf("%s(%d,%d)",append?"#APPEND":"#SET",returnToken.pos.lineno,returnToken.pos.column);
+    ECLlocation location = returnToken.pos;
+    const char * directive = append ? "#APPEND" : "#SET";
 
     IIdAtom * name = NULL;
     if (yyLex(returnToken, LEXnone, 0) != '(')
@@ -884,11 +873,11 @@ void HqlLex::doSet(attribute & returnToken, bool append)
     }
 
     StringBuffer curParam("(");
-    if (getParameter(curParam,forwhat.str()))
+    if (getParameter(curParam, directive, location))
     {
         reportError(returnToken, ERR_OPERANDS_TOOMANY, "Too many operands");
         StringBuffer dummy;
-        while (getParameter(dummy,forwhat.str()))
+        while (getParameter(dummy, directive, location))
             ;
     }
     curParam.append(')');
@@ -904,8 +893,9 @@ void HqlLex::doSet(attribute & returnToken, bool append)
 
 void HqlLex::doLine(attribute & returnToken)
 {
-    StringBuffer forwhat;
-    forwhat.appendf("LINE(%d,%d)",returnToken.pos.lineno,returnToken.pos.column);
+    ECLlocation location = returnToken.pos;
+    const char * directive = "#LINE";
+
     if (yyLex(returnToken, LEXnone, 0) != '(')
     {
         reportError(returnToken, ERR_EXPECTED_LEFTCURLY, "( expected");
@@ -913,7 +903,7 @@ void HqlLex::doLine(attribute & returnToken)
         return;
     }
     StringBuffer curParam("(");
-    bool moreParams = getParameter(curParam, forwhat.str());
+    bool moreParams = getParameter(curParam, directive, location);
     curParam.append(')');
 
     IValue *value = parseConstExpression(returnToken.pos, curParam, queryTopXmlScope());
@@ -930,11 +920,11 @@ void HqlLex::doLine(attribute & returnToken)
         ECLlocation extrapos;
         getPosition(extrapos);
 
-        if (getParameter(curParam,forwhat.str()))
+        if (getParameter(curParam, directive, location))
         {
             reportError(extrapos, ERR_OPERANDS_TOOMANY, "Too many operands");
             StringBuffer dummy;
-            while (getParameter(dummy,forwhat.str()))
+            while (getParameter(dummy, directive, location))
                 ;
         }
         curParam.append(')');
@@ -1016,8 +1006,8 @@ void HqlLex::doSlashSlashHash(attribute const & returnToken, const char * comman
 
 void HqlLex::doError(attribute & returnToken, bool isError)
 {
-    StringBuffer forwhat;
-    forwhat.appendf("%s(%d,%d)",isError?"#ERROR":"#WARNING",returnToken.pos.lineno,returnToken.pos.column);
+    ECLlocation location = returnToken.pos;
+    const char * directive = isError ? "#ERROR" : "#WARNING";
 
     if (yyLex(returnToken, LEXnone, 0) != '(')
     {
@@ -1026,11 +1016,11 @@ void HqlLex::doError(attribute & returnToken, bool isError)
         return;
     }
     StringBuffer curParam("(");
-    if (getParameter(curParam,forwhat.str()))
+    if (getParameter(curParam, directive, location))
     {
         reportError(returnToken, ERR_OPERANDS_TOOMANY, "Too many operands");
         StringBuffer dummy;
-        while (getParameter(dummy,forwhat.str()))
+        while (getParameter(dummy, directive, location))
             ;
     }
     curParam.append(')');
@@ -1050,8 +1040,8 @@ void HqlLex::doError(attribute & returnToken, bool isError)
 
 void HqlLex::doExport(attribute & returnToken, bool toXml)
 {
-    StringBuffer forwhat;
-    forwhat.appendf("#EXPORT(%d,%d)",returnToken.pos.lineno,returnToken.pos.column);
+    ECLlocation location = returnToken.pos;
+    const char * directive = "#EXPORT";
 
     IIdAtom * exportname = NULL;
     if (yyLex(returnToken, LEXnone, 0) != '(')
@@ -1076,7 +1066,7 @@ void HqlLex::doExport(attribute & returnToken, bool toXml)
     for (;;)
     {
         StringBuffer curParam("SIZEOF(");
-        bool more = getParameter(curParam,"#EXPORT");
+        bool more = getParameter(curParam, directive, location);
         curParam.append(",MAX)");
 
         OwnedHqlExpr expr;
@@ -1120,8 +1110,8 @@ void HqlLex::doExport(attribute & returnToken, bool toXml)
 
 void HqlLex::doTrace(attribute & returnToken)
 {
-    StringBuffer forwhat;
-    forwhat.appendf("#TRACE(%d,%d)",returnToken.pos.lineno,returnToken.pos.column);
+    ECLlocation location = returnToken.pos;
+    const char * directive = "#TRACE";
 
     if (yyLex(returnToken, LEXnone, 0) != '(')
     {
@@ -1131,11 +1121,11 @@ void HqlLex::doTrace(attribute & returnToken)
     }
     StringBuffer curParam("(");
 
-    if (getParameter(curParam,forwhat.str()))
+    if (getParameter(curParam, directive, location))
     {
         reportError(returnToken, ERR_OPERANDS_TOOMANY, "Too many operands");
         StringBuffer dummy;
-        while (getParameter(dummy,forwhat.str()))
+        while (getParameter(dummy, directive, location))
             ;
     }
     curParam.append(')');
@@ -1155,10 +1145,8 @@ void HqlLex::doTrace(attribute & returnToken)
 
 void HqlLex::doFor(attribute & returnToken, bool doAll)
 {
-    //MTIME_SECTION(timer, "HqlLex::doFor")
-
-    StringBuffer forwhat;
-    forwhat.appendf("#FOR(%d,%d)",returnToken.pos.lineno,returnToken.pos.column);
+    ECLlocation location = returnToken.pos;
+    const char * directive = "#FOR";
 
     IIdAtom * name = NULL;
     if (yyLex(returnToken, LEXnone, 0) != '(')
@@ -1182,7 +1170,7 @@ void HqlLex::doFor(attribute & returnToken, bool doAll)
     if (tok == '(')
     {
         forFilterText.append('(');
-        while (getParameter(forFilterText, forwhat.str()))
+        while (getParameter(forFilterText, directive, location))
             forFilterText.append(") AND (");
         forFilterText.append(')');
         tok = yyLex(returnToken, LEXnone, 0);
@@ -1205,7 +1193,7 @@ void HqlLex::doFor(attribute & returnToken, bool doAll)
 
         if (tok == EOF)
         {
-            reportError(returnToken, ERR_TMPLT_HASHENDEXPECTED, "EOF encountered inside %s: #END expected", forwhat.str());
+            reportError(location, ERR_TMPLT_HASHENDEXPECTED, "EOF encountered inside %s: #END expected", directive);
             clearNestedHash();      // prevent unnecessary more error messages
             return;
         }
@@ -1230,11 +1218,8 @@ void HqlLex::doFor(attribute & returnToken, bool doAll)
 
 void HqlLex::doLoop(attribute & returnToken)
 {
-    ECLlocation loopPos = returnToken.pos;
-
-    StringBuffer forwhat;
-    forwhat.appendf("#LOOP(%d,%d)",loopPos.lineno,loopPos.column);
-
+    ECLlocation location = returnToken.pos;
+    const char * directive = "#LOOP";
 
     // Now gather the tokens we are going to repeat...
     StringBuffer forBodyText;
@@ -1247,7 +1232,7 @@ void HqlLex::doLoop(attribute & returnToken)
 
         if (tok == EOF)
         {
-            reportError(returnToken, ERR_TMPLT_HASHENDEXPECTED, "EOF encountered inside %s: #END expected",forwhat.str());
+            reportError(location, ERR_TMPLT_HASHENDEXPECTED, "EOF encountered inside #LOOP: #END expected");
             clearNestedHash();      // prevent unnecessary more error messages
             return;
         }
@@ -1259,12 +1244,12 @@ void HqlLex::doLoop(attribute & returnToken)
     }
     if (!hasHashbreak)
     {
-        reportError(returnToken, ERR_TMPLT_NOBREAKINLOOP,"No #BREAK inside %s: infinite loop will occur", forwhat.str());
+        reportError(location, ERR_TMPLT_NOBREAKINLOOP, "No #BREAK inside %s: infinite loop will occur", directive);
         return;
     }
 
     ::Release(forLoop);
-    forLocation = loopPos;
+    forLocation = location;
     forLoop = new CDummyScopeIterator(ensureTopXmlScope());
     forFilter.clear();
     forBody.setown(createFileContentsFromText(forBodyText, sourcePath, yyParser->inSignedModule, yyParser->gpgSignature, 0));
@@ -1275,15 +1260,18 @@ void HqlLex::doLoop(attribute & returnToken)
 
 void HqlLex::doGetDataType(attribute & returnToken)
 {
+    ECLlocation location = returnToken.pos;
+    const char * directive = "#GETDATATYPE";
+
     int tok = yyLex(returnToken, LEXnone, 0);
     if (tok != '(')
         reportError(returnToken, ERR_EXPECTED_LEFTCURLY, "( expected"); // MORE - make it fatal!
     StringBuffer curParam("(");
-    if (getParameter(curParam, "#GETDATATYPE"))
+    if (getParameter(curParam, directive, location))
     {
         reportError(returnToken, ERR_OPERANDS_TOOMANY, "too many operands");
         StringBuffer dummy;
-        while (getParameter(dummy,"#GETDATATYPE"))
+        while (getParameter(dummy, directive, location))
             ;
     }
     curParam.append(')');
@@ -1310,8 +1298,8 @@ StringBuffer& HqlLex::doGetDataType(StringBuffer & type, const char * text, int 
 
 int HqlLex::doHashText(attribute & returnToken)
 {
-    StringBuffer forwhat;
-    forwhat.appendf("#TEXT(%d,%d)",returnToken.pos.lineno,returnToken.pos.column);
+    ECLlocation location = returnToken.pos;
+    const char * directive = "#TEXT";
 
     if (yyLex(returnToken, LEXnone, 0) != '(')
     {
@@ -1322,7 +1310,7 @@ int HqlLex::doHashText(attribute & returnToken)
     }
 
     StringBuffer parameterText;
-    bool moreParams = getParameter(parameterText, forwhat.str());
+    bool moreParams = getParameter(parameterText, directive, location);
     if (!moreParams)
     {
         while (parameterText.length() && parameterText.charAt(0)==' ')
@@ -1332,7 +1320,7 @@ int HqlLex::doHashText(attribute & returnToken)
     {
         reportError(returnToken, ERR_OPERANDS_TOOMANY, "Too many operands");
         StringBuffer dummy;
-        while (getParameter(dummy,forwhat.str()))
+        while (getParameter(dummy, directive, location))
             ;
     }
 
@@ -1343,24 +1331,24 @@ int HqlLex::doHashText(attribute & returnToken)
 
 void HqlLex::doInModule(attribute & returnToken)
 {
-#ifdef TIMING_DEBUG
-    MTIME_SECTION(timer, "HqlLex::doInModule");
-#endif
+    ECLlocation location = returnToken.pos;
+    const char * directive = "#INMODULE";
+
     int tok = yyLex(returnToken, LEXnone, 0);
     if (tok != '(')
         reportError(returnToken, ERR_EXPECTED_LEFTCURLY, "( expected");
 
     StringBuffer moduleName, attrName;
 
-    if (getParameter(moduleName,"#INMODULE"))
+    if (getParameter(moduleName, directive, location))
     {
-        if (getParameter(attrName,"#INMODULE"))
+        if (getParameter(attrName, directive, location))
         {
             reportError(returnToken, ERR_OPERANDS_TOOMANY, "too many operands");
 
             /* skip the rest */
             StringBuffer dummy;
-            while (getParameter(dummy,"#INMODULE"))
+            while (getParameter(dummy, directive, location))
                 ;
         }
     }
@@ -1488,15 +1476,18 @@ void HqlLex::declareUniqueName(const char *name, const char * pattern)
 
 void HqlLex::doIsValid(attribute & returnToken)
 {
+    ECLlocation location = returnToken.pos;
+    const char * directive = "#ISVALID";
+
     int tok = yyLex(returnToken, LEXnone, 0);
     if (tok != '(')
         reportError(returnToken, ERR_EXPECTED_LEFTCURLY, "( expected");
     StringBuffer curParam("(");
-    if (getParameter(curParam,"#ISVALID"))
+    if (getParameter(curParam, directive, location))
     {
         reportError(returnToken, ERR_OPERANDS_TOOMANY, "too many operands");
         StringBuffer dummy;
-        while (getParameter(dummy,"#ISVALID"))
+        while (getParameter(dummy, directive, location))
             ;
     }
     curParam.append(')');
@@ -1634,7 +1625,7 @@ void HqlLex::doPreprocessorLookup(const attribute & errpos, bool stringify, int 
 
 
 //Read the text of a parameter, but also have a good guess at whether it is defined.
-bool HqlLex::getDefinedParameter(StringBuffer &curParam, attribute & returnToken, const char* for_what, SharedHqlExpr & resolved)
+bool HqlLex::getDefinedParameter(StringBuffer &curParam, attribute & returnToken, const char* directive, const ECLlocation & location, SharedHqlExpr & resolved)
 {
     enum { StateStart, StateDot, StateSelectId, StateFailed } state = StateStart;
     unsigned parenDepth = 1;
@@ -1669,9 +1660,7 @@ bool HqlLex::getDefinedParameter(StringBuffer &curParam, attribute & returnToken
             break;
         case EOF:
             {
-                StringBuffer msg("EOF encountered while gathering parameters for ");
-                msg.append(for_what);
-                reportError(returnToken, ERR_TMPLT_EOFINPARAM, "%s", msg.str());
+                reportError(location, ERR_TMPLT_EOFINPARAM, "EOF encountered while gathering parameters for %s", directive);
             }
             return false;
         case UNKNOWN_ID:
@@ -1728,15 +1717,15 @@ bool HqlLex::getDefinedParameter(StringBuffer &curParam, attribute & returnToken
 
 bool HqlLex::doIsDefined(attribute & returnToken)
 {
-    StringBuffer forwhat;
-    forwhat.appendf("#ISDEFINED(%d,%d)",returnToken.pos.lineno,returnToken.pos.column);
+    ECLlocation location = returnToken.pos;
+    const char * directive = "#ISDEFINED";
 
     if (!assertNextOpenBra())
         return false;
 
     OwnedHqlExpr resolved;
     StringBuffer paramText;
-    bool hasMore = getDefinedParameter(paramText, returnToken, forwhat.str(), resolved);
+    bool hasMore = getDefinedParameter(paramText, returnToken, directive, location, resolved);
     if (hasMore)
         reportError(returnToken, ERR_EXPECTED, "Expected ')'");
     return resolved != NULL;
@@ -1745,8 +1734,8 @@ bool HqlLex::doIsDefined(attribute & returnToken)
 
 void HqlLex::doDefined(attribute & returnToken)
 {
-    StringBuffer forwhat;
-    forwhat.appendf("#DEFINED(%d,%d)",returnToken.pos.lineno,returnToken.pos.column);
+    ECLlocation location = returnToken.pos;
+    const char * directive = "#DEFINED";
 
     if (!assertNextOpenBra())
         return;
@@ -1754,9 +1743,9 @@ void HqlLex::doDefined(attribute & returnToken)
     OwnedHqlExpr resolved;
     StringBuffer param1Text;
     StringBuffer param2Text;
-    bool hasMore = getDefinedParameter(param1Text, returnToken, forwhat.str(), resolved);
+    bool hasMore = getDefinedParameter(param1Text, returnToken, directive, location, resolved);
     if (hasMore)
-        hasMore = getParameter(param2Text, forwhat.str());
+        hasMore = getParameter(param2Text, directive, location);
     if (hasMore)
         reportError(returnToken, ERR_EXPECTED, "Expected ')'");
 
@@ -1854,17 +1843,20 @@ int hexchar(char c)
 
 void HqlLex::doApply(attribute & returnToken)
 {
+    ECLlocation location = returnToken.pos;
+    const char * directive = "#APPLY";
+
     int tok = yyLex(returnToken, LEXnone, 0);
     if (tok != '(')
         reportError(returnToken, ERR_EXPECTED_LEFTCURLY, "( expected"); // MORE - make it fatal!
 
     int line = returnToken.pos.lineno, col = returnToken.pos.column;
     StringBuffer curParam("(");
-    if (getParameter(curParam, "#APPLY"))
+    if (getParameter(curParam, directive, location))
     {
         reportError(returnToken, ERR_OPERANDS_TOOMANY, "too many operands");
         StringBuffer dummy;
-        while (getParameter(dummy, "#APPLY"))
+        while (getParameter(dummy, directive, location))
             ;
     }
     curParam.append(')');
@@ -1879,17 +1871,20 @@ void HqlLex::doApply(attribute & returnToken)
 
 void HqlLex::doMangle(attribute & returnToken, bool de)
 {
+    ECLlocation location = returnToken.pos;
+    const char * directive = de?"#DEMANGLE":"#MANGLE";
+
     int tok = yyLex(returnToken, LEXnone, 0);
     if (tok != '(')
         reportError(returnToken, ERR_EXPECTED_LEFTCURLY, "( expected"); // MORE - make it fatal!
 
     ECLlocation pos = returnToken.pos;
     StringBuffer curParam("(");
-    if (getParameter(curParam, de?"#DEMANGLE":"#MANGLE"))
+    if (getParameter(curParam, directive, location))
     {
         reportError(returnToken, ERR_OPERANDS_TOOMANY, "too many operands");
         StringBuffer dummy;
-        while (getParameter(dummy, de?"#DEMANGLE":"MANGLE"))
+        while (getParameter(dummy, directive, location))
             ;
     }
     curParam.append(')');

